crates/ruff/src/message/diff.rs

```diff
@@ -0,0 +1,194 @@
+                        Line {
+                            index: old_index,
+                            width: digit_with
+                        },
+                        Line {
+                            index: new_index,
+                            width: digit_with
+                        },
+                        line_style.apply_to(sign).bold()
+                    )?;
+
+                    for (emphasized, value) in change.iter_strings_lossy() {
+                        if emphasized {
+                            write!(f, "{}", line_style.apply_to(&value).underline().on_black())?;
+                        } else {
+                            write!(f, "{}", line_style.apply_to(&value))?;
+                        }
+                    }
+                    if change.missing_newline() {
+                        writeln!(f)?;
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+struct LineStyle {
+    fgcolor: Option<Color>,
+    style: Option<Styles>,
+}
+
+impl LineStyle {
+    fn apply_to(&self, input: &str) -> ColoredString {
+        let mut colored = ColoredString::from(input);
+        if let Some(color) = self.fgcolor {
+            colored = colored.color(color);
+        }
+
+        if let Some(style) = self.style {
+            match style {
+                Styles::Clear => colored.clear(),
+                Styles::Bold => colored.bold(),
+                Styles::Dimmed => colored.dimmed(),
+                Styles::Underline => colored.underline(),
+                Styles::Reversed => colored.reversed(),
+                Styles::Italic => colored.italic(),
+                Styles::Blink => colored.blink(),
+                Styles::Hidden => colored.hidden(),
+                Styles::Strikethrough => colored.strikethrough(),
+            }
+        } else {
+            colored
+        }
+    }
+}
+
+impl From<ChangeTag> for LineStyle {
+    fn from(value: ChangeTag) -> Self {
+        match value {
+            ChangeTag::Equal => LineStyle {
+                fgcolor: None,
+                style: Some(Styles::Dimmed),
+            },
+            ChangeTag::Delete => LineStyle {
+                fgcolor: Some(Color::Red),
+                style: None,
+            },
+            ChangeTag::Insert => LineStyle {
+                fgcolor: Some(Color::Green),
+                style: None,
+            },
+        }
+    }
+}
+
+struct Line {
+    index: Option<OneIndexed>,
+    width: NonZeroUsize,
+}
+
+impl Display for Line {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        match self.index {
+            None => {
+                for _ in 0..self.width.get() {
+                    f.write_str(" ")?;
+                }
+                Ok(())
+            }
+            Some(idx) => write!(f, "{:<width$}", idx, width = self.width.get()),
+        }
+    }
+}
+
+/// Calculate the length of the string representation of `value`
```

Comment:
Not sure if it was removed but we do have `num_digits` in `printer.rs`, maybe whatever was relying on that can leverage this now too.